src/crimson/osd/main.cc

```diff
@@ -56,7 +56,12 @@ int main(int argc, char* argv[])
         return sharded_perf_coll().stop();
       });
       seastar::engine().at_exit([&] {
```

Comment:
when with --mkfs, can't stop these messengers since it is only initialized in OSD::start, --mkfs not execute OSD::start.

Reply:
i think we have two more options

- to check before calling stop. namely
```c++
  if (state.is_initializing()) {
    return seastar::now();
  }
  return gate.close().then([this] {
  /// ...
  }
```
- to hook `stop()` where `OSD::start()` is called.

IMHO, the second one is better. as it is the caller `OSD::start()` is aware of if `OSD::start()` is called. the callee is not responsible for covering the oversight of caller.

Reply:
@liu-chunmei i added "crimson/osd: call engine().exit(0) after mkfs" in #26697 to partially address this issue. but as you and Yingxin pointed out, we have a more complicated problem regarding to the ordering of sharded service shutdown. i am adding a task in crimson project at https://github.com/ceph/ceph/projects/2#card-18842255 to track it.